rotkehlchen/chain/evm/decoding/open_ocean/decoder.py

```diff
@@ -0,0 +1,188 @@
+        )
+        self._decode_swap(
+            transaction=context.transaction,
+            decoded_events=context.decoded_events,
+            spend_asset=spend_asset,
+            spend_amount=spend_amount,
+            receive_asset=receive_asset,
+            receive_amount=receive_amount,
+        )
+        return DEFAULT_DECODING_OUTPUT
+
+    @staticmethod
+    def _decode_swap(
+            transaction: EvmTransaction,
+            decoded_events: list['EvmEvent'],
+            spend_asset: 'Asset | None' = None,
+            spend_amount: 'FVal | None' = None,
+            receive_asset: 'Asset | None' = None,
+            receive_amount: 'FVal | None' = None,
+    ) -> None:
+        """Decode OpenOcean swap in/out events.
+        If assets and amounts are None they are ignored, and only the event type is matched.
+        Checks TRADE event types in addition to the plain SPEND/RECEIVE since an event may
+        have already been processed by a previous decoder like uniswap.
+        Returns the list of decoded events.
+        """
+        in_event, out_event = None, None
+        for event in decoded_events:
+            if ((
+                (event.event_type == HistoryEventType.SPEND and event.event_subtype == HistoryEventSubType.NONE) or  # noqa: E501
+                (event.event_type == HistoryEventType.TRADE and event.event_subtype == HistoryEventSubType.SPEND)) and  # noqa: E501
+                (spend_asset is None or event.asset == spend_asset) and
+                (spend_amount is None or event.balance.amount == spend_amount)
+            ):
+                event.event_type = HistoryEventType.TRADE
+                event.event_subtype = HistoryEventSubType.SPEND
+                event.notes = f'Swap {event.balance.amount} {event.asset.resolve_to_asset_with_symbol().symbol} in {OPENOCEAN_LABEL}'  # noqa: E501
+                event.counterparty = CPT_OPENOCEAN
+                out_event = event
+            elif ((
+                (event.event_type == HistoryEventType.RECEIVE and event.event_subtype == HistoryEventSubType.NONE) or  # noqa: E501
+                (event.event_type == HistoryEventType.TRADE and event.event_subtype == HistoryEventSubType.RECEIVE)) and  # noqa: E501
+                (receive_asset is None or event.asset == receive_asset) and
+                (receive_amount is None or event.balance.amount == receive_amount)
+            ):
+                event.event_type = HistoryEventType.TRADE
+                event.event_subtype = HistoryEventSubType.RECEIVE
+                event.notes = f'Receive {event.balance.amount} {event.asset.resolve_to_asset_with_symbol().symbol} from {OPENOCEAN_LABEL} swap'  # noqa: E501
+                event.counterparty = CPT_OPENOCEAN
+                in_event = event
+
+        if in_event is None or out_event is None:
+            log.error(f'Failed to find both out and in events for {OPENOCEAN_LABEL} swap transaction {transaction}')  # noqa: E501
```

Comment:
no need to reshufle here?

Reply:
No - if one of the events isn't found there's no reshuffling that can be done is there?

BTW, I changed this error to only be a warning, since in rare cases this does happen, for instance when meta mask uses openocean to do a swap.